### PR TITLE
[SR-7297][TypeChecker] Discarding a function from a discardableResult method causes a compiler error

### PR DESCRIPTION
**Pull request by chuganzy**
From the ticket:

This example doesn't compile:

```swift
struct S {
  @discardableResult
  func foo() -> () -> String { fatalError() }
}

let s = S()
s.foo() // error: Expression resolves to an unused function
```

But really it should.

Resolves [SR-7297](https://bugs.swift.org/browse/SR-7297).

### CODE REVIEW

**Comment by hamishknight** [comment]
Thanks for fixing this! Would it be possible to re-use the later logic that finds callees and disregards those marked as `@discardableResult` for this case?

https://github.com/apple/swift/blob/e1f50b2d367ee565a935a5d36957bdf73cf02640/lib/Sema/TypeCheckStmt.cpp#L1206-L1240

It seems to me that we're trying to emit the "unused function expression" diagnostic too early, we should first wait to discount applies of callees that are `@discardableResult` (i.e we should be looking to emit the diagnostic after this logic). That way we can get rid of the `isDiscardable` variable and avoid duplicating the logic to find the callee.

File: test/expr/expressions.swift [hunk 1]
```diff
@@ -702,6 +707,12 @@ func unusedExpressionResults() {
   let optionalc:C? = nil
   optionalc?.method()  // ok
   optionalc?.method  // expected-error {{expression resolves to an unused function}}
+
+  optionalc?.method2 // expected-error {{expression resolves to an unused function}}
+  optionalc?.method2() // expected-error {{expression resolves to an unused function}}
+
+  optionalc?.method3 // ok
```
Comment (jrose-apple) [inline test/expr/expressions.swift]: This one isn't right; this *is* an unused function, not an unused result.

**Review by jrose-apple — COMMENTED** [review]
Thanks for picking this up! I think Hamish's suggestion is a good idea, but I wanted to point out an incorrect test you've added.

File: test/expr/expressions.swift (hunk 1 above)
Comment (chuganzy) [reply test/expr/expressions.swift]: Oh? This currently (am using v4.1) passes tho? Is it a bug then?

Comment (jrose-apple) [reply test/expr/expressions.swift]: Yes, that's a bug too! Whoops.

Comment (jrose-apple) [reply test/expr/expressions.swift]: Representing methods as `(A) -> () -> String` has so many unfortunate implications…!

Comment (chuganzy) [reply test/expr/expressions.swift]: Let me confirm so everything gets clear before fixing.

```swift
final class A {
  func method1() -> String {
    return ""
  }
}

// Expected behaviours 
A.method1 // expression resolves to an unused function
A().method1 // expression resolves to an unused function


```

Comment (chuganzy) [reply test/expr/expressions.swift]: Let me confirm so everything gets clear before fixing:

```swift
final class A {
  func method1() -> String {
    return ""
  }

  @discardableResult
  func method2() -> String {
    return ""
  }

  @discardableResult
  func method3() -> () -> String {
    return { "" }
  }
}


// Expected
A.method1   // expression resolves to an unused function
A().method1 // expression resolves to an unused function
A.method2   // expression resolves to an unused function
A.method3   // expression resolves to an unused function

// Bugs
A().method2 // passes -> expression resolves to an unused function
A().method3() // expression resolves to an unused function -> passes
```

Comment (jrose-apple) [reply test/expr/expressions.swift]: Yes, that looks correct to me. If it's hard to fix the `A.methodN(A())` cases it's probably okay to have those pass even if they really shouldn't; doing that is rare anyway.

Comment (chuganzy) [reply test/expr/expressions.swift]: New question - what do you think of `A().method3()()`? I feel it should complain about unused result but it does nothing so far..

Comment (jrose-apple) [reply test/expr/expressions.swift]: I agree that that one should complain too! It seems rare as well, though.

**Comment by rudkx** [comment]
@chuganzy It wasn't clear to me the last time I opened this PR whether it was ready to run tests on and review, or if there was something more you wanted to do for this.

Should I start the tests and take a look?

**Comment by chuganzy** [comment]
Sorry that I leave this open so long but this is not ready to go..
To be honest, I recently broke my wrist's bone and am not able to work on this for a while, so it might take more to wrap this up..

**Comment by rudkx** [comment]
@chuganzy No worries!

Heal up and I'll look forward to future updates.

**Comment by jrose-apple** [comment]
Ah, @chuganzy, are you still interested in this? If not one of us can take it over.

**Comment by sindresorhus** [comment]
@chuganzy ⬆️ Would be good with an answer either way so this PR can continue. The issue is a pretty annoying limitation when you need to return an optional closure.

**Comment by CodaFi** [comment]
Closing this due to age and inactivity.  Thanks for the patch @chuganzy, it will be very valuable to a future contributor.  We've reset the assignee on the linked SR.

**Pull request closed without merging**